pkg/profiler/cpu/cpu_test.go

```diff
@@ -15,25 +15,352 @@
 package cpu
 
 import (
+	"context"
+	"fmt"
+	"os/exec"
+	"strconv"
+	"strings"
 	"syscall"
 	"testing"
+	"time"
 	"unsafe"
 
 	bpf "github.com/aquasecurity/libbpfgo"
+	"github.com/go-kit/log"
+	"github.com/google/pprof/profile"
 	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/procfs"
+	"github.com/prometheus/prometheus/model/relabel"
 	"github.com/stretchr/testify/require"
 
+	"github.com/parca-dev/parca-agent/pkg/address"
+	"github.com/parca-dev/parca-agent/pkg/ksym"
 	"github.com/parca-dev/parca-agent/pkg/logger"
+	"github.com/parca-dev/parca-agent/pkg/metadata"
+	"github.com/parca-dev/parca-agent/pkg/metadata/labels"
+	"github.com/parca-dev/parca-agent/pkg/objectfile"
+	"github.com/parca-dev/parca-agent/pkg/perf"
+	"github.com/parca-dev/parca-agent/pkg/process"
+	"github.com/parca-dev/parca-agent/pkg/profiler"
+	"github.com/parca-dev/parca-agent/pkg/symbol"
+	"github.com/parca-dev/parca-agent/pkg/vdso"
 )
 
+type Sample struct {
+	labels  model.LabelSet
+	profile *profile.Profile
+}
+
+type TestProfileWriter struct {
+	samples []Sample
+}
+
+func NewTestProfileWriter() *TestProfileWriter {
+	return &TestProfileWriter{samples: make([]Sample, 0)}
+}
+
+func (tpw *TestProfileWriter) Write(_ context.Context, labels model.LabelSet, profile *profile.Profile) error {
+	tpw.samples = append(tpw.samples, Sample{
+		labels:  labels,
+		profile: profile,
+	})
+	return nil
+}
+
+// SampleForProcess returns the firt or last matching sample for a given
+// PID.
+func (tpw *TestProfileWriter) SampleForProcess(pid int, last bool) *Sample {
+	for i := range tpw.samples {
+		var sample Sample
+		if last {
+			sample = tpw.samples[len(tpw.samples)-1-i]
+		} else {
+			sample = tpw.samples[i]
+		}
+
+		foundPid, err := strconv.Atoi(string(sample.labels["pid"]))
+		if err != nil {
+			panic("label pid is not a valid integer")
+		}
+
+		if foundPid == pid {
+			return &sample
+		}
+	}
+
+	return nil
+}
+
+type LocalSymbolizer struct {
+	addr2line string
+	timeout   time.Duration
+}
+
+func NewLocalSymbolizer() *LocalSymbolizer {
+	return &LocalSymbolizer{addr2line: "/usr/bin/addr2line", timeout: time.Second * 5}
```

Comment:
Just thinking out loud here: Maybe we can use the upstream parca symbolizer as well. However, this could potentially not an ideal thing if our symbolizer has issues.

Reply:
Left some notes in the description about this. When I originally started implementing what you mention, didn't want to add one more Parca dependency. Wanted to keep this as separate as possible.

Something I have been meaning to do but I haven't had bandwidth yet (no pun intended!!) is to move this "local" symbolizer to a proper package, so we can use it in Parca to diff the results of our symbolizer in Go vs standard CLI tools such as `addr2line` and others to ensure our code work as expected.